File: cronografo/cronografo/ViewModels/MainViewModel.cs
using System;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Diagnostics;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using System.Windows.Input;
using cronografo.Models;
using cronografo.Services;
using Plugin.BLE.Abstractions.Contracts;

namespace cronografo.ViewModels
{
    public class MainViewModel : INotifyPropertyChanged
    {
        private readonly IBluetoothLEService _bleService;
        private string _statusText = "Desconectado";
        private bool _isScanning;
        private bool _isConnected;
        private string _pelletWeightText = "0.52";
        private string _lastShotNumber = "-";
        private string _lastShotVelocity = "0.0";
        private string _lastShotEnergy = "0.0";

        private string _avgVelocity = "0.0";
        private string _minVelocity = "0.0";
        private string _maxVelocity = "0.0";
        private string _avgEnergy = "0.0";
        private string _minEnergy = "0.0";
        private string _maxEnergy = "0.0";

        private string _receivedBuffer = "";
        
        // Estado del parser
        private int? _currentParsingNumber;
        private float? _currentParsingVelocity;
        private float? _currentParsingEnergy;

        public event PropertyChangedEventHandler? PropertyChanged;

        public ObservableCollection<IDevice> Devices { get; } = new();
        public ObservableCollection<ShotModel> Shots { get; } = new();

        public string StatusText
        {
            get => _statusText;
            set => SetProperty(ref _statusText, value);
        }

        public bool IsScanning
        {
            get => _isScanning;
            set => SetProperty(ref _isScanning, value);
        }

        public bool IsConnected
        {
            get => _isConnected;
            set => SetProperty(ref _isConnected, value);
        }

        public string PelletWeightText
        {
            get => _pelletWeightText;
            set => SetProperty(ref _pelletWeightText, value);
        }

        public string LastShotNumber
        {
            get => _lastShotNumber;
            set => SetProperty(ref _lastShotNumber, value);
        }

        public string LastShotVelocity
        {
            get => _lastShotVelocity;
            set => SetProperty(ref _lastShotVelocity, value);
        }

        public string LastShotEnergy
        {
            get => _lastShotEnergy;
            set => SetProperty(ref _lastShotEnergy, value);
        }

        public string AvgVelocity
        {
            get => _avgVelocity;
            set => SetProperty(ref _avgVelocity, value);
        }

        public string MinVelocity
        {
            get => _minVelocity;
            set => SetProperty(ref _minVelocity, value);
        }

        public string MaxVelocity
        {
            get => _maxVelocity;
            set => SetProperty(ref _maxVelocity, value);
        }

        public string AvgEnergy
        {
            get => _avgEnergy;
            set => SetProperty(ref _avgEnergy, value);
        }

        public string MinEnergy
        {
            get => _minEnergy;
            set => SetProperty(ref _minEnergy, value);
        }

        public string MaxEnergy
        {
            get => _maxEnergy;
            set => SetProperty(ref _maxEnergy, value);
        }

        public ICommand ScanCommand { get; }
        public ICommand ConnectCommand { get; }
        public ICommand DisconnectCommand { get; }
        public ICommand UpdateWeightCommand { get; }
        public ICommand ResetCommand { get; }

        public MainViewModel(IBluetoothLEService bleService)
        {
            _bleService = bleService;
            _bleService.DataReceived += OnDataReceived;

            ScanCommand = new Command(async () => await ExecuteScanCommand());
            ConnectCommand = new Command<IDevice>(async (device) => await ExecuteConnectCommand(device));
            DisconnectCommand = new Command(async () => await ExecuteDisconnectCommand());
            UpdateWeightCommand = new Command(async () => await ExecuteUpdateWeightCommand());
            ResetCommand = new Command(async () => await ExecuteResetCommand());
        }

        private async Task ExecuteScanCommand()
        {
            if (IsScanning) return;

            IsScanning = true;
            StatusText = "Escaneando dispositivos...";
            Devices.Clear();

            try
            {
                await _bleService.StartScanningAsync((device) =>
                {
                    // Filtrar dispositivos sin nombre o duplicados
                    if (!string.IsNullOrEmpty(device.Name) && Devices.All(d => d.Id != device.Id))
                    {
                        // Priorizar CronoABC
                        if (device.Name.Equals("CronoABC", StringComparison.OrdinalIgnoreCase))
                        {
                            Devices.Insert(0, device);
                        }
                        else
                        {
                            Devices.Add(device);
                        }
                    }
                });
            }
            catch (Exception ex)
            {
                StatusText = $"Error al escanear: {ex.Message}";
            }
            finally
            {
                IsScanning = false;
                if (!IsConnected)
                {
                    StatusText = "Búsqueda finalizada.";
                }
            }
        }

        private async Task ExecuteConnectCommand(IDevice device)
        {
            if (device == null) return;

            StatusText = $"Conectando a {device.Name}...";
            await _bleService.StopScanningAsync();

            bool success = await _bleService.ConnectDeviceAsync(device);
            if (success)
            {
                IsConnected = true;
                StatusText = $"Conectado a {device.Name}";
                
                // Cargar estadísticas iniciales o peso
                await Task.Delay(500);
                await _bleService.SendCommandAsync("peso");
            }
            else
            {
                IsConnected = false;
                StatusText = "Error al intentar conectar.";
            }
        }

        private async Task ExecuteDisconnectCommand()
        {
            StatusText = "Desconectando...";
            await _bleService.DisconnectDeviceAsync();
            IsConnected = false;
            StatusText = "Desconectado";
        }

        private async Task ExecuteUpdateWeightCommand()
        {
            if (!IsConnected) return;

            if (float.TryParse(PelletWeightText, out float weight) && weight > 0)
            {
                await _bleService.SendCommandAsync($"peso {weight:F2}");
                StatusText = $"Peso actualizado a {weight:F2}g en el hardware.";
            }
            else
            {
                StatusText = "Masa inválida. Introduce un número mayor a 0.";
            }
        }

        private async Task ExecuteResetCommand()
        {
            if (IsConnected)
            {
                await _bleService.SendCommandAsync("reset");
            }

            Shots.Clear();
            RecalculateStatistics();
            LastShotNumber = "-";
            LastShotVelocity = "0.0";
            LastShotEnergy = "0.0";
            StatusText = "Sesión reiniciada.";
        }

        private void OnDataReceived(string text)
        {
            _receivedBuffer += text;
            ProcessBuffer();
        }

        private void ProcessBuffer()
        {
            while (true)
            {
                int newlineIdx = _receivedBuffer.IndexOf('\n');
                if (newlineIdx == -1) break;

                string line = _receivedBuffer.Substring(0, newlineIdx).Trim();
                _receivedBuffer = _receivedBuffer.Substring(newlineIdx + 1);

                if (string.IsNullOrEmpty(line)) continue;

                ParseLine(line);
            }
        }

        private void ParseLine(string line)
        {
            // Procesamiento de comandos informativos devueltos
            if (line.StartsWith("Nuevo peso:", StringComparison.OrdinalIgnoreCase) || 
                line.StartsWith("Peso actual:", StringComparison.OrdinalIgnoreCase))
            {
                // Extraer el peso del mensaje para actualizar la UI
                var match = Regex.Match(line, @"\d+(\.\d+)?");
                if (match.Success)
                {
                    PelletWeightText = match.Value;
                }
                return;
            }

            if (line.StartsWith("Estadísticas reiniciadas", StringComparison.OrdinalIgnoreCase))
            {
                return;
            }

            // Parsing de disparos
            // Ejemplo:
            // Disparo #1
            //  250.5 m/s
            //  16.3 Jul
            if (line.StartsWith("Disparo #", StringComparison.OrdinalIgnoreCase))
            {
                if (int.TryParse(line.Substring(9), out int num))
                {
                    _currentParsingNumber = num;
                }
            }
            else if (line.EndsWith("m/s", StringComparison.OrdinalIgnoreCase))
            {
                string valStr = line.Replace("m/s", "", StringComparison.OrdinalIgnoreCase).Trim();
                if (float.TryParse(valStr, out float vel))
                {
                    _currentParsingVelocity = vel;
                }
            }
            else if (line.EndsWith("Jul", StringComparison.OrdinalIgnoreCase))
            {
                string valStr = line.Replace("Jul", "", StringComparison.OrdinalIgnoreCase).Trim();
                if (float.TryParse(valStr, out float nrg))
                {
                    _currentParsingEnergy = nrg;
                }
            }

            // Si ya tenemos los tres componentes del disparo actual, guardarlo e informar
            if (_currentParsingNumber.HasValue && _currentParsingVelocity.HasValue && _currentParsingEnergy.HasValue)
            {
                var newShot = new ShotModel
                {
                    Number = _currentParsingNumber.Value,
                    Velocity = _currentParsingVelocity.Value,
                    Energy = _currentParsingEnergy.Value
                };

                // Insertar al inicio de la lista para ver el más nuevo primero
                MainThread.BeginInvokeOnMainThread(() =>
                {
                    Shots.Insert(0, newShot);
                    
                    LastShotNumber = $"Disparo #{newShot.Number}";
                    LastShotVelocity = $"{newShot.Velocity:F1}";
                    LastShotEnergy = $"{newShot.Energy:F1}";
                    
                    RecalculateStatistics();

                    // Anuncio por voz
                    AnnounceShot(newShot);
                });

                // Resetear estado del parser
                _currentParsingNumber = null;
                _currentParsingVelocity = null;
                _currentParsingEnergy = null;
            }
        }

        private void AnnounceShot(ShotModel shot)
        {
            // Decir velocidad en formato amigable
            // Reemplazar el punto decimal por "punto" para una pronunciación natural en español
            string velocidadPronunciar = shot.Velocity.ToString("F1").Replace(",", " punto ").Replace(".", " punto ");
            string mensaje = $"{shot.Velocity:F0} metros por segundo";
            
            // Usar TextToSpeech de MAUI
            Task.Run(async () =>
            {
                try
                {
                    await TextToSpeech.Default.SpeakAsync(mensaje, new SpeechOptions
                    {
                        Volume = 1.0f,
                        Locale = (await TextToSpeech.Default.GetLocalesAsync())
                                    .FirstOrDefault(l => l.Language.StartsWith("es", StringComparison.OrdinalIgnoreCase))
                    });
                }
                catch (Exception ex)
                {
                    Debug.WriteLine($"Error de TextToSpeech: {ex.Message}");
                }
            });
        }

        private void RecalculateStatistics()
        {
            if (Shots.Count == 0)
            {
                AvgVelocity = "0.0";
                MinVelocity = "0.0";
                MaxVelocity = "0.0";
                AvgEnergy = "0.0";
                MinEnergy = "0.0";
                MaxEnergy = "0.0";
                return;
            }

            var velocities = Shots.Select(s => s.Velocity).ToList();
            var energies = Shots.Select(s => s.Energy).ToList();

            AvgVelocity = $"{velocities.Average():F1}";
            MinVelocity = $"{velocities.Min():F1}";
            MaxVelocity = $"{velocities.Max():F1}";

            AvgEnergy = $"{energies.Average():F1}";
            MinEnergy = $"{energies.Min():F1}";
            MaxEnergy = $"{energies.Max():F1}";
        }

        protected bool SetProperty<T>(ref T storage, T value, [CallerMemberName] string propertyName = "")
        {
            if (Equals(storage, value)) return false;
            storage = value;
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
            return true;
        }
    }
}

File: cronografo_android/app/src/main/java/com/luis/cronografo/MainViewModel.kt
package com.luis.cronografo

import android.bluetooth.BluetoothDevice
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.luis.cronografo.ble.BleManager
import com.luis.cronografo.tts.VoiceAnnouncer
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import java.util.Locale

class MainViewModel(
    val bleManager: BleManager,
    private val voiceAnnouncer: VoiceAnnouncer
) : ViewModel() {

    private val _isScanning = MutableStateFlow(false)
    val isScanning: StateFlow<Boolean> = _isScanning.asStateFlow()

    private val _currentLanguage = MutableStateFlow("ca")
    val currentLanguage: StateFlow<String> = _currentLanguage.asStateFlow()

    private val _statusText = MutableStateFlow<UiText>(UiText.StringResource(R.string.status_disconnected))
    val statusText: StateFlow<UiText> = _statusText.asStateFlow()

    val devices = bleManager.devices
    val isConnected = bleManager.isConnected

    private val _pelletWeightText = MutableStateFlow("0.52")
    val pelletWeightText: StateFlow<String> = _pelletWeightText.asStateFlow()

    private val _shots = MutableStateFlow<List<ShotModel>>(emptyList())
    val shots: StateFlow<List<ShotModel>> = _shots.asStateFlow()

    private val _lastShotNumber = MutableStateFlow<UiText>(UiText.DynamicString("-"))
    val lastShotNumber: StateFlow<UiText> = _lastShotNumber.asStateFlow()

    private val _lastShotVelocity = MutableStateFlow("0.0")
    val lastShotVelocity: StateFlow<String> = _lastShotVelocity.asStateFlow()

    private val _lastShotEnergy = MutableStateFlow("0.0")
    val lastShotEnergy: StateFlow<String> = _lastShotEnergy.asStateFlow()

    private val _avgVelocity = MutableStateFlow("0.0")
    val avgVelocity: StateFlow<String> = _avgVelocity.asStateFlow()
    private val _minVelocity = MutableStateFlow("0.0")
    val minVelocity: StateFlow<String> = _minVelocity.asStateFlow()
    private val _maxVelocity = MutableStateFlow("0.0")
    val maxVelocity: StateFlow<String> = _maxVelocity.asStateFlow()

    private val _avgEnergy = MutableStateFlow("0.0")
    val avgEnergy: StateFlow<String> = _avgEnergy.asStateFlow()
    private val _minEnergy = MutableStateFlow("0.0")
    val minEnergy: StateFlow<String> = _minEnergy.asStateFlow()
    private val _maxEnergy = MutableStateFlow("0.0")
    val maxEnergy: StateFlow<String> = _maxEnergy.asStateFlow()

    private var receivedBuffer = ""
    private var currentParsingNumber: Int? = null
    private var currentParsingVelocity: Float? = null
    private var currentParsingEnergy: Float? = null

    init {
        viewModelScope.launch {
            bleManager.incomingData.collectLatest { data ->
                receivedBuffer += data
                processBuffer()
            }
        }
    }

    fun startScan() {
        _isScanning.value = true
        _statusText.value = UiText.StringResource(R.string.status_scanning)
        bleManager.startScan()
        
        // Stop scan after 10s
        viewModelScope.launch {
            kotlinx.coroutines.delay(10000)
            if (_isScanning.value) {
                stopScan()
            }
        }
    }

    fun stopScan() {
        _isScanning.value = false
        bleManager.stopScan()
        if (!isConnected.value) {
            _statusText.value = UiText.StringResource(R.string.status_scan_finished)
        }
    }

    fun connect(device: BluetoothDevice) {
        receivedBuffer = ""
        currentParsingNumber = null
        currentParsingVelocity = null
        currentParsingEnergy = null

        val deviceName = device.name ?: "Dispositivo"
        _statusText.value = UiText.StringResource(R.string.status_connecting, deviceName)
        bleManager.connect(device)
        _statusText.value = UiText.StringResource(R.string.status_connected, deviceName)
        
        viewModelScope.launch {
            kotlinx.coroutines.delay(1000)
            bleManager.sendCommand("peso")
        }
    }

    fun disconnect() {
        bleManager.disconnect()
        _statusText.value = UiText.StringResource(R.string.status_disconnected)
    }

    fun setLanguage(langCode: String) {
        _currentLanguage.value = langCode
        voiceAnnouncer.setLanguage(langCode)
    }

    fun updateWeightText(weight: String) {
        _pelletWeightText.value = weight
    }

    fun sendWeight() {
        val weight = _pelletWeightText.value.toFloatOrNull()
        if (weight != null && weight > 0) {
            bleManager.sendCommand("peso ${String.format(Locale.US, "%.2f", weight)}")
            _statusText.value = UiText.StringResource(R.string.status_weight_updated, weight.toString())
        }
    }

    fun resetSession() {
        receivedBuffer = ""
        currentParsingNumber = null
        currentParsingVelocity = null
        currentParsingEnergy = null
        
        bleManager.sendCommand("reset")
        _shots.value = emptyList()
        recalculateStatistics()
        _lastShotNumber.value = UiText.DynamicString("-")
        _lastShotVelocity.value = "0.0"
        _lastShotEnergy.value = "0.0"
        _statusText.value = UiText.StringResource(R.string.status_session_reset)
    }

    private fun processBuffer() {
        // Reemplazar \r con \n para manejar saltos de línea irregulares del hardware
        receivedBuffer = receivedBuffer.replace('\r', '\n')
        
        while (true) {
            val newlineIdx = receivedBuffer.indexOf('\n')
            if (newlineIdx == -1) break
            
            val line = receivedBuffer.substring(0, newlineIdx).trim()
            receivedBuffer = receivedBuffer.substring(newlineIdx + 1)
            
            if (line.isNotEmpty()) {
                // Si el hardware concatenó "Disparo #" con un texto anterior (ej. por faltar un \n)
                var cleanLine = line
                val disparoIdx = cleanLine.indexOf("Disparo #", ignoreCase = true)
                if (disparoIdx > 0) {
                    val beforeDisparo = cleanLine.substring(0, disparoIdx).trim()
                    if (beforeDisparo.isNotEmpty()) {
                        parseLine(beforeDisparo)
                    }
                    cleanLine = cleanLine.substring(disparoIdx).trim()
                }
                
                parseLine(cleanLine)
            }
        }
    }

    private fun parseLine(line: String) {
        if (line.startsWith("Nuevo peso:", ignoreCase = true) || 
            line.startsWith("Peso actual:", ignoreCase = true)) {
            val regex = Regex("""\d+(\.\d+)?""")
            val match = regex.find(line)
            if (match != null) {
                _pelletWeightText.value = match.value
            }
            return
        }

        if (line.startsWith("Disparo #", ignoreCase = true)) {
            val numStr = line.substring(9)
            currentParsingNumber = numStr.toIntOrNull()
        } else if (line.endsWith("m/s", ignoreCase = true)) {
            val valStr = line.replace("m/s", "", ignoreCase = true).trim()
            currentParsingVelocity = valStr.toFloatOrNull()
        } else if (line.endsWith("Jul", ignoreCase = true)) {
            val valStr = line.replace("Jul", "", ignoreCase = true).trim()
            currentParsingEnergy = valStr.toFloatOrNull()
        }

        if (currentParsingNumber != null && currentParsingVelocity != null && currentParsingEnergy != null) {
            val newShot = ShotModel(
                number = currentParsingNumber!!,
                velocity = currentParsingVelocity!!,
                energy = currentParsingEnergy!!
            )
            
            val currentList = _shots.value.toMutableList()
            currentList.add(0, newShot)
            _shots.value = currentList
            
            _lastShotNumber.value = UiText.StringResource(R.string.shot_number, newShot.number.toString())
            _lastShotVelocity.value = String.format(Locale.US, "%.1f", newShot.velocity)
            _lastShotEnergy.value = String.format(Locale.US, "%.1f", newShot.energy)
            
            recalculateStatistics()
            voiceAnnouncer.announceShot(newShot.velocity, newShot.energy)

            currentParsingNumber = null
            currentParsingVelocity = null
            currentParsingEnergy = null
        }
    }

    private fun recalculateStatistics() {
        val shotsList = _shots.value
        if (shotsList.isEmpty()) {
            _avgVelocity.value = "0.0"
            _minVelocity.value = "0.0"
            _maxVelocity.value = "0.0"
            _avgEnergy.value = "0.0"
            _minEnergy.value = "0.0"
            _maxEnergy.value = "0.0"
            return
        }

        val velocities = shotsList.map { it.velocity }
        val energies = shotsList.map { it.energy }

        _avgVelocity.value = String.format(Locale.US, "%.1f", velocities.average())
        _minVelocity.value = String.format(Locale.US, "%.1f", velocities.minOrNull() ?: 0f)
        _maxVelocity.value = String.format(Locale.US, "%.1f", velocities.maxOrNull() ?: 0f)

        _avgEnergy.value = String.format(Locale.US, "%.1f", energies.average())
        _minEnergy.value = String.format(Locale.US, "%.1f", energies.minOrNull() ?: 0f)
        _maxEnergy.value = String.format(Locale.US, "%.1f", energies.maxOrNull() ?: 0f)
    }

    override fun onCleared() {
        super.onCleared()
        voiceAnnouncer.shutdown()
        bleManager.disconnect()
    }
}
